url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/network/datacenter/NetworkCloudletSpaceSharedScheduler.java

```diff
@@ -0,0 +1,854 @@
+import org.cloudbus.cloudsim.core.CloudSim;
+import org.cloudbus.cloudsim.core.CloudSimTags;
+
+/**
+ * CloudletSchedulerSpaceShared implements a policy of scheduling performed by a virtual machine. It
+ * consider that there will be only one cloudlet per VM. Other cloudlets will be in a waiting list.
+ * We consider that file transfer from cloudlets waiting happens before cloudlet execution. I.e.,
+ * even though cloudlets must wait for CPU, data transfer happens as soon as cloudlets are
+ * submitted.
+ * 
+ * @author Saurabh Kumar Garg
+ * @author Saurabh Kumar Garg
+ * @since CloudSim Toolkit 3.0
+ */
+public class NetworkCloudletSpaceSharedScheduler extends CloudletScheduler {
+
+	/** The cloudlet waiting list. */
+	private List<? extends ResCloudlet> cloudletWaitingList;
+
+	/** The cloudlet exec list. */
+	private List<? extends ResCloudlet> cloudletExecList;
+
+	/** The cloudlet paused list. */
+	private List<? extends ResCloudlet> cloudletPausedList;
+
+	/** The cloudlet finished list. */
+	private List<? extends ResCloudlet> cloudletFinishedList;
+
+	/** The current CPUs. */
+	protected int currentCpus;
+
+	/** The used PEs. */
+	protected int usedPes;
+        
+        private Datacenter datacenter;
+	// for network
+
+	public Map<Integer, List<HostPacket>> pkttosend;
+
+	public Map<Integer, List<HostPacket>> pktrecv;
+
+	/**
+	 * Creates a new CloudletSchedulerSpaceShared object. This method must be invoked before
+	 * starting the actual simulation.
+	 * 
+	 * @pre $none
+	 * @post $none
+	 */
+	public NetworkCloudletSpaceSharedScheduler() {
+		super();
+		cloudletWaitingList = new ArrayList<ResCloudlet>();
+		cloudletExecList = new ArrayList<ResCloudlet>();
+		cloudletPausedList = new ArrayList<ResCloudlet>();
+		cloudletFinishedList = new ArrayList<ResCloudlet>();
+		usedPes = 0;
+		currentCpus = 0;
+		pkttosend = new HashMap<Integer, List<HostPacket>>();
+		pktrecv = new HashMap<Integer, List<HostPacket>>();
+	}
+        
+        public NetworkCloudletSpaceSharedScheduler(Datacenter datacenter_) {
+		super();
+		cloudletWaitingList = new ArrayList<ResCloudlet>();
+		cloudletExecList = new ArrayList<ResCloudlet>();
+		cloudletPausedList = new ArrayList<ResCloudlet>();
+		cloudletFinishedList = new ArrayList<ResCloudlet>();
+		usedPes = 0;
+		currentCpus = 0;
+		pkttosend = new HashMap<Integer, List<HostPacket>>();
+		pktrecv = new HashMap<Integer, List<HostPacket>>();
+                datacenter=datacenter_;
+	}
+
+	/**
+	 * Updates the processing of cloudlets running under management of this scheduler.
+	 * 
+	 * @param currentTime current simulation time
+	 * @param mipsShare array with MIPS share of each processor available to the scheduler
+	 * @return time predicted completion time of the earliest finishing cloudlet, or 0 if there is
+	 *         no next events
+	 * @pre currentTime >= 0
+	 * @post $none
+	 */
+	@Override
+	public double updateVmProcessing(double currentTime, List<Double> mipsShare) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 25. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 121-205, lines 214-248, lines 251-263 into separate methods.